plasmapy/utils/code_repr.py

```diff
@@ -0,0 +1,321 @@
+
+    is_ndarray = isinstance(arg.value, np.ndarray)
+    formatted = _code_repr_of_ndarray(arg.value) if is_ndarray else repr(arg.value)
+
+    if arg.unit == u.dimensionless_unscaled:
+        formatted += "*u.dimensionless_unscaled"
+    else:
+        for base, power in zip(arg.unit.bases, arg.unit.powers):
+            if power == -1:
+                formatted += f"/u.{base}"
+            elif power == 1:
+                formatted += f"*u.{base}"
+            else:
+                formatted += f"*u.{base}**{power}"
+
+    return formatted
+
+
+def _code_repr_of_arg(arg) -> str:
+    """Transform an argument into a format as would appear in a function call."""
+    if hasattr(arg, "__name__"):
+        return arg.__name__
+
+    function_for_type = {
+        u.Quantity: _code_repr_of_quantity,
+        np.ndarray: _code_repr_of_ndarray,
+    }
+
+    for arg_type in function_for_type.keys():
+        if isinstance(arg, arg_type):
+            code_repr_func = function_for_type[arg_type]
+            return code_repr_func(arg)
+
+    return repr(arg)
+
+
+def _code_repr_of_args_and_kwargs(args: Any = tuple(), kwargs: Dict = {}) -> str:
+    """
+    Take positional and keyword arguments, and format them into a
+    string as they would appear in a function call (excluding parentheses).
+    """
+    args_collection = args if isinstance(args, tuple) else (args,)
+
+    args_and_kwargs = ""
+
+    for arg in args_collection:
+        args_and_kwargs += f"{_code_repr_of_arg(arg)}, "
+
+    for kwarg in kwargs.keys():
+        args_and_kwargs += f"{kwarg}={_code_repr_of_arg(kwargs[kwarg])}, "
+
+    if args_and_kwargs[-2:] == ", ":
+        args_and_kwargs = args_and_kwargs[:-2]
+
+    return args_and_kwargs
+
+
+def _name_with_article(ex: Exception) -> str:
+    """
+    Return a string with an indefinite article and name of
+    exception ``ex``.
+
+    Notes
+    -----
+    If this function is to be expanded for cases beyond exceptions,
+    we would need to either expand the treatment of cases that do not
+    follow the general rule, or use a library like `inflect` on PyPI.
+    """
+    starts_with_vowel_but_uses_a = ["use", "uni"]
+    name = ex.__name__
+    use_an = all(
+        [
+            name[0] in "aeiouAEIOU",
+            name[0:3].lower() not in starts_with_vowel_but_uses_a,
+        ]
+    )
+    indefinite_article = "an" if use_an else "a"
+    return f"{indefinite_article} {name}"
+
+
+def _substitute_module_shortcuts(module_name):
+    """Substitute common import shortcuts within module names."""
+    replacements = {
+        "numpy": "np",
+        "builtins": "",
+        "astropy.units.core": "u",
+        "astropy.units.quantity": "u",
+        "astropy.units.decorators": "u",
+        "astropy.units": "u",
+    }
+    for old, new in replacements.items():
+        if module_name.startswith(old):
+            module_name = module_name.replace(old, new, 1)
+    return module_name
```

Comment:
If `_substitute_module_shortcuts()` is only going to be used by `_object_name()` then it's code should be contained with it `_object_name()`; otherwise, `_substitute_module_shortcuts()` should have its own dedicated tests.  If neither of those approaches are taken then these two functions are not 100% independent of each other can lead to problems down the line.